File: 234.Palindrome_linked_list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head) -> bool:
        
        ## BRUTE FORCE - beats 72%, extra space
        ## T(N): O(N)
        ## SPACE COMPLEXITY: O(N), extra array to store reversed order
        ## CONVERT TO ARRAY
        ## REVERESE IT AND COMPARE
        
        arr = []
        
        # convert to array
        while head:
            arr.append(head.val)
            head = head.next

        # reverse array
        # rev_arr = []
        # rev_arr = arr[::-1]

        l, r = 0, len(arr) -1
        while l <= r:
            if arr[l] != arr[r]:
                return False
            l += 1
            r -= 1

        return True

        # if arr == rev_arr:
        #     return True
        
    def isPalindrome2(self, head) -> bool:
        
        # T(N) = O(N)
        # SPACE - O(1)

        # FIND MIDDLE TO
        # REVERSE SECOND HALF 
        # POINT TO FIRST AND LAST AND CHECK FOR PALINDROME

        # find middle using floyd's cycle detection algorithm
        # also finds middle if no cycle (fast and slow pointers)
        slow, fast = head, head

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        # reverse second half
        prev = None
        while slow:                         # slow points to null
            temp = slow.next
            slow.next = prev
            prev = slow
            slow = temp
    
        # second half of LL is reversed
        # check palindrome
        left, right = head, prev
        while right:                        # until right reaches the middle i.e. None
            if left.val != right.val:
                return False
            left = left.next
            right = right.next
        return True

        

File: leetcode_java/234. Palindrome Linked List.java
package leetcode_java;


//  Definition for singly-linked list.
 class ListNode {
     int val;
     ListNode next;
     ListNode() {}
     ListNode(int val) { this.val = val; }
     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 }
 

class Solution {

    public static ListNode findMiddle(ListNode node) {
        ListNode fast = node, slow = node;

        // fast will go till last node or till just right of last node
        // while ((fast.next != null) && (fast != null)) { this won't work bcz what if fast is null, you're checking null.next
        while ((fast != null) && (fast.next != null)) {
            slow = slow.next;
            fast = fast.next.next;
        }
        return slow;  // slow will point to the middle
    }

    public static ListNode reverseLL(ListNode node) {
        ListNode prev = null, curr = node, nxt = null;

        while (curr != null){
            nxt = curr.next;
            curr.next = prev;
            prev = curr;
            curr = nxt;
        }
        return prev;    // prev will now point to the start of reversed list (end)

    }


    public boolean isPalindrome(ListNode head) {
        // find middle using fast and slow
        ListNode middle = findMiddle(head);

        // reverse the second half
        ListNode tail = reverseLL(middle);

        // check palindrome
        ListNode start = head, end = tail;
        while (end != null) {
            if (start.val != end.val) {
                return false;
            }
            start = start.next;
            end = end.next;
        }
        return true;
        
    }
}